core/src/main/java/jenkins/scm/RunWithSCMMixIn.java

```diff
@@ -0,0 +1,161 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.scm;
+
+import hudson.model.AbstractBuild;
+import hudson.model.AbstractProject;
+import hudson.model.Job;
+import hudson.model.Queue;
+import hudson.model.Result;
+import hudson.model.Run;
+import hudson.model.User;
+import hudson.scm.ChangeLogSet;
+import hudson.util.AdaptedIterator;
+import jenkins.triggers.SCMTriggerItem;
+import jenkins.util.SystemProperties;
+import org.kohsuke.stapler.export.Exported;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.AbstractSet;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * @since FIXME
+ */
+public abstract class RunWithSCMMixIn<JobT extends Job<JobT, RunT> & Queue.Task,
+        RunT extends Run<JobT, RunT> & RunWithSCMMixIn.RunWithSCM<JobT, RunT> & Queue.Executable> {
+    /**
+     * Set if we want the blame information to flow from upstream to downstream build.
+     */
+    private static final boolean upstreamCulprits = SystemProperties.getBoolean("hudson.upstreamCulprits");
+
+    protected abstract RunT asRun();
+
+    public abstract List<ChangeLogSet<? extends ChangeLogSet.Entry>> getChangeSets();
+
+    /**
+     * List of users who committed a change since the last non-broken build till now.
+     *
+     * <p>
+     * This list at least always include people who made changes in this build, but
+     * if the previous build was a failure it also includes the culprit list from there.
+     *
+     * @return
+     *      can be empty but never null.
+     */
+    @Exported
+    @Nonnull public Set<User> getCulprits() {
+        if (asRun().getCulpritIds() == null) {
+            Set<User> r = new HashSet<User>();
+            RunT p = asRun().getPreviousCompletedBuild();
+            if (p != null && asRun().isBuilding()) {
+                Result pr = p.getResult();
+                if (pr != null && pr.isWorseThan(Result.SUCCESS)) {
+                    // we are still building, so this is just the current latest information,
+                    // but we seems to be failing so far, so inherit culprits from the previous build.
+                    // isBuilding() check is to avoid recursion when loading data from old Hudson, which doesn't record
+                    // this information
+                    r.addAll(p.getCulprits());
+                }
+            }
+            for (ChangeLogSet<? extends ChangeLogSet.Entry> c : getChangeSets()) {
+                for (ChangeLogSet.Entry e : c)
+                    r.add(e.getAuthor());
+            }
+
+            if (p instanceof AbstractBuild && upstreamCulprits) {
```

Comment:
I suppose you are deliberately punting in this PR on dealing with `DependencyGraph`.